### PR TITLE
Handle spilled registers

### PR DESCRIPTION
**Pull request by JHodorHodor**

### CODE REVIEW

File: app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt [hunk 1]
```diff
@@ -1,20 +1,150 @@
 package compiler.intermediate_form
 
+import compiler.common.intermediate_form.Allocation
+
 object SpillsHandlingAllocation {
 
-    data class AllocationResultWithSpilledOffset(
+    data class Result(
```
Comment (woocash2) [inline app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:7]: Good idea, this long name is not needed

File: app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt [hunk 2]
```diff
@@ -1,20 +1,150 @@
...
+
+            reservedRegistersNumber ++
+        }
+    }
+
+    data class SpilledInstruction(
+        val instruction: Instruction,
+        val spilledUsedRegisters: Set<Register>,
+        val spilledDefinedRegisters: Set<Register>,
+        val numberOfSpilledRegisters: Int = (spilledUsedRegisters + spilledDefinedRegisters).size,
```
Comment (woocash2) [inline app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:67]: Does this `+` work like a union of sets?

Comment (JHodorHodor) [reply app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:67]: yes

File: app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt [hunk 3]
```diff
@@ -1,20 +1,150 @@
...
+    ): Map<Register, ULong> {
+        val spillsPossibleColors = spilledRegisters.map { Register() }
+        val spillsColoring: Map<Register, Register> = allocator.allocateRegisters(
+            Liveness.LivenessGraphs(
+                Liveness.inducedSubgraph(livenessGraphs.interferenceGraph, spilledRegisters),
+                Liveness.inducedSubgraph(livenessGraphs.copyGraph, spilledRegisters),
+            ),
+            spillsPossibleColors,
+        ).allocatedRegisters
+        val spillsColors = spillsPossibleColors.filter { it in spillsColoring.values }
```
Comment (woocash2) [inline app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:93]: Isn't it just `spollsColoring.values`?

File: app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt [hunk 4]
```diff
@@ -1,20 +1,150 @@
...
+
+    private fun transformProgram(
+        linearProgram: List<Asmable>,
+        instructionsToSpilled: Map<Int, SpilledInstruction>,
+        allocatedRegisters: Map<Register, Register>,
+        spillsColoring: Map<Register, ULong>,
+    ): List<Asmable> {
+        val newLinearProgram = mutableListOf<Asmable>()
+
+        fun handleSpilledInstruction(spilledInstruction: SpilledInstruction) {
```
Comment (woocash2) [inline app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:118]: Is moving in this function for sure correct?
Now we save each used register to memory before usage and override register with memory after definition.
Shouldn't we save to memory before definition and load from memory before usage

**Review by woocash2 — APPROVED** [review]
Looks good, accepted but take a look at one comment on `handleSpilledInstruction` to make sure it's correct.

File: app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt (hunk 4 above)
Comment (JHodorHodor) [reply app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:118]: Hmm, first, we load to registers to be used from memory (MoveRM), then we perform the instruction and finally, we save defined registers to memory (MoverMR). How can we save to memory before definition (when the thing is not defined yet)?

Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:118]: Oh, I see now. `RM` means memory -> register and `MR` register -> memory. I thought it was the other way around.

Comment (JHodorHodor) [reply app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:118]: I hope I'm right
```
MoveRM(val reg_dest: Register, val mem_src: Addressing)
MoveMR(val mem_dest: Addressing, val reg_src: Register)
```

Comment (woocash2) [reply app/src/main/kotlin/compiler/intermediate_form/SpillsHandlingAllocation.kt:118]: Yeah, all good.

**Pull request merged**